applications/services/storage/storage_processing.c

```diff
@@ -260,6 +267,141 @@ static bool storage_process_file_eof(Storage* app, File* file) {
+    // cursor needs to be at the beginning of file
+    uint8_t magic_buf[encryption_magic_size];
+    storage_process_file_read(app, file, &magic_buf, encryption_magic_size);
+    if(memcmp(encryption_magic_bytes, magic_buf, encryption_magic_size) != 0) {
+        // file is not encrypted, move cursor back and stop
+        storage_process_file_seek(app, file, -1 * encryption_magic_size, false);
+        return FSE_OK;
+    }
+
+    // read key slot id and initialization vector
+    uint8_t key_slot;
+    uint8_t iv[ENCRYPTION_IV_SIZE];
+    storage_file_read(file, &key_slot, sizeof(key_slot));
+    storage_file_read(file, &iv, ENCRYPTION_IV_SIZE);
+
+    // store encryption info
+    file->encryption = storage_file_encryption_alloc(key_slot);
+
+    // decrypt file
+    if(!furi_hal_crypto_store_load_key(key_slot, iv)) {
+        return FSE_INTERNAL;
+    }
+    // we will decrypt file in-place, so we need two cursors:
+    // decryption cursor - points to next encrypted chunk of data, always in front of data cursor
+    uint64_t dec_cursor = storage_process_file_tell(app, file);
+    // decrypted data cursor - points to the end of decrypted data
+    uint64_t data_cursor = 0;
+
+    uint8_t chunk_size = 64;
+    uint8_t crypt_buf[chunk_size];
+    uint8_t dec_buf[chunk_size];
+    while(!storage_process_file_eof(app, file)) {
+        storage_process_file_seek(app, file, dec_cursor, true);
+        uint16_t read = storage_process_file_read(app, file, &crypt_buf, chunk_size);
+        dec_cursor += read;
+        furi_hal_crypto_decrypt(crypt_buf, (uint8_t*)&dec_buf, read);
+        storage_process_file_seek(app, file, data_cursor, true);
+        data_cursor += storage_process_file_write(app, file, dec_buf, read);
+    }
+    furi_hal_crypto_store_unload_key(key_slot);
+
+    // truncate exceeding data and reset cursor
+    storage_process_file_seek(app, file, data_cursor, true);
+    storage_process_file_truncate(app, file);
+    storage_process_file_seek(app, file, 0, true);
+
+    file->encryption->decrypted = true;
+    return FSE_OK;
+}
+
+static FS_Error
+    storage_process_file_encrypt(Storage* app, const char* path, const uint8_t key_slot) {
+    FS_Error error = FSE_OK;
+
+    Stream* stream_from = file_stream_alloc(app);
+    Stream* stream_to = file_stream_alloc(app);
+
+    const char* enc_filepath = strcat((char*)path, ENCRYPTION_EXT);
+
+    if(file_stream_open(stream_from, path, FSAM_READ, FSOM_OPEN_EXISTING) &&
+       file_stream_open(stream_to, enc_filepath, FSAM_WRITE, FSOM_CREATE_NEW)) {
+        // check free space
+        uint64_t total_space;
+        uint64_t free_space;
+        storage_common_fs_info(app, path, &total_space, &free_space);
+
+        if(free_space > stream_size(stream_from) + encryption_header_size) {
+            // create new random initialization vector
+            uint8_t iv[ENCRYPTION_IV_SIZE];
+            srand(DWT->CYCCNT);
+            furi_hal_random_fill_buf(iv, ENCRYPTION_IV_SIZE);
```

Comment:
IV is generated here using srand PRNG + CYCCNT (i.e. cycle counter) as seed,
* Only 32 bits of "entropy" for CYCCNT [link](https://developer.arm.com/documentation/ddi0403/d/Debug-Architecture/ARMv7-M-Debug/The-Data-Watchpoint-and-Trace-unit/CYCCNT-cycle-counter-and-related-timers?lang=en)
* Predicted (CYCCNT is just incrementing counter), making predicted IV attacks possible (like complete plaintext recovery for low entropy plaintexts) as the resulted encryption is no longer CPA secure 
* srand is not photographically secure PRG ([even with better seed](https://crypto.stackexchange.com/questions/52000/how-can-c-rand-be-exploited-if-a-secure-seed-is-used) as it has only 31 bit inner state)

Original flipper firmware already implemented much better random generation using sts32 [hardware random generator](https://www.st.com/content/ccc/resource/training/technical/product_training/group0/3e/47/41/d7/09/ea/42/3b/STM32WB-Security-Random-Number-Generator-RNG/files/STM32WB-Security-Random-Number-Generator-RNG.pdf/_jcr_content/translations/en.STM32WB-Security-Random-Number-Generator-RNG.pdf). See [furi_hal_random.c](https://github.com/DarkFlippers/unleashed-firmware/blob/78c64737aa014a83920767235ae90c5d983713f2/firmware/targets/f7/furi_hal/furi_hal_random.c#L59-L61) `rand()` function.